lusidtools/cocoon/cocoon_printer.py

```diff
@@ -0,0 +1,171 @@
+    :param list required_keys: List of required keys
+    :return:
+    """
+    for key in required_keys:
+        if key not in target_dict.keys():
+            raise ValueError(f"'{key}' missing from {target_name}. {target_name} must include {required_keys}"
+                             f"as keys")
+
+
+def get_errors_from_response(list_of_API_exceptions):
+    """
+    This function gets the status code and reason from API exception
+    :param list list_of_API_exceptions:
+    :return pd.DataFrame: a Pandas DataFrame containing the reasons and status codes for ApiExceptions from a request
+    """
+    # get API exception
+
+    status = [item.status for item in list_of_API_exceptions]
+    reason = [item.reason for item in list_of_API_exceptions]
+    items_error = np.array([reason, status]).T.tolist()
+
+    return pd.DataFrame(items_error, columns=["error_items", "status"])
+
+def backtracking_search_for_substring(search_string: str, index: int, flag: str)-> int:
+    """
+    This function performs a recursive backtracking search for a substring preceded by an identifying string
+    :param search_string:
+    :param index: index to begin search
+    :return:
+    """
+    return index if search_string[index-1:index-1+len(flag)] == flag else backtracking_search_for_substring(search_string, index - 1, flag)
+
+
+def get_portfolio_from_href(href, file_type):
+    """
+    This function finds the protfolio code contained within a href for a given file_type
+    :param list href: Full href from LUSID response
+    :param str file_type:
+    :return list: portfolio codes taken from hrefs
+    """
+    flag = "/"
+    for item in href:
+        if f"/{file_type}?" not in item:
+            raise ValueError(f"Invalid href structure: '/{file_type}?' does not exist in href")
+
+    codes = [item[backtracking_search_for_substring(search_string=item,
+                                                    index=item.find(f"/{file_type}?")-1,
+                                                    flag=flag)
+                  :item.find(f"/{file_type}?")]
+             for item in href]
+    return codes
+
+
+def format_instruments_response(response: dict) -> (
+        pd.DataFrame, pd.DataFrame, pd.DataFrame):
+    """
+    This function unpacks a response from instrument requests and returns successful, failed and errored statuses for
+    request constituents.
+    :param dict response: response from Lusid-python-tools
+    :return: pd.DataFrame: Successful calls from request
+    :return: pd.DataFrame: Error responses from request that fail (APIExceptions: 400 errors)
+    :return: pd.DataFrame: Failed responses that LUSID rejected
+    """
+    file_type = "instruments"
+    check_dict_for_required_keys(response[file_type], f"Response from {file_type} request", ["errors", "success"])
+
+
+    # get success and failures
+    items_success = [batch.values[key].name for batch in response[file_type]["success"] for key in batch.values.keys()]
```

Comment:
Use the key on the response here instead of name which is guaranteed to be unique.